### PR TITLE
CLN: NDFrame._where

### LINKED ISSUE #38729: BUG: DataFrame.where with EA other
```
df = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
arr = pd.array([7, pd.NA, 9])
mask = np.ones(df.shape, dtype=bool)

>>> result = df.where(mask, arr)
AttributeError: 'IntegerArray' object has no attribute 'reshape'
```

we can kludge around this by using `np.reshape(arr, shape)` instead of `arr.reshape(shape)`, but that'll cast to object, which is not great.

Non-kludge solution: 2D EAs

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #38729 
- [ ] tests added / passed
- [ ] passes `black pandas`
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/generic.py [hunk 1]
```diff
@@ -8825,22 +8829,39 @@ def _where(
...
 
+                elif other.ndim < self.ndim:
+                    # TODO(EA2D): avoid object-dtype cast in EA case GH#38729
+                    other = other._values
+                    if axis == 0:
+                        other = np.reshape(other, (-1, 1))
+                    if axis == 1:
+                        other = np.reshape(other, (1, -1))
+
+                    other = np.broadcast_to(other, self.shape)
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py:8852]: When does this happen? When `other` is a Series, I think? 
In that case you can also broadcast the Series to an aligned DataFrame without the conversion to numpy?

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/generic.py:8852]: Also, does this fix a case? Does it need a test?

Comment (jbrockmendel) [reply pandas/core/generic.py:8852]: > When other is a Series, I think?

Yes.

> In that case you can also broadcast the Series to an aligned DataFrame without the conversion to numpy?

Most ways of doing that would make copies on DataFrame construction, which this avoids.

> Also, does this fix a case? Does it need a test?

Yes, fixes #38729, will add tests.

File: pandas/core/generic.py [hunk 2]
```diff
@@ -8786,6 +8787,9 @@ def _where(
         """
         inplace = validate_bool_kwarg(inplace, "inplace")
 
+        if axis is not None:
```
Comment (jreback) [inline pandas/core/generic.py:8790]: we do this i am pretty sure in the caller, so why do you need it here? (e.g. when is this ever non-None)?

File: pandas/core/generic.py [hunk 3]
```diff
@@ -8825,22 +8829,39 @@ def _where(
...
-                ):
+                if axis is None and not other._indexed_same(self):
                     raise InvalidIndexError
 
+                elif other.ndim < self.ndim:
+                    # TODO(EA2D): avoid object-dtype cast in EA case GH#38729
+                    other = other._values
+                    if axis == 0:
+                        other = np.reshape(other, (-1, 1))
+                    if axis == 1:
```
Comment (jreback) [inline pandas/core/generic.py:8849]: elif

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/generic.py:8790]: for starters the default is None, and no axis is passed when this is called from `Series.__setitem__`

File: pandas/core/generic.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/generic.py:8849]: sure

File: pandas/tests/frame/indexing/test_where.py [hunk 4]
```diff
@@ -653,3 +653,22 @@ def test_where_categorical_filtering(self):
         expected.loc[0, :] = np.nan
 
         tm.assert_equal(result, expected)
+
```
Comment (jreback) [inline pandas/tests/frame/indexing/test_where.py:656]: can you move these to the generic EA indexing tests

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/indexing/test_where.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/tests/frame/indexing/test_where.py:656]: this file is the canonical place for DataFrame.where tests

Comment (jreback) [reply pandas/tests/frame/indexing/test_where.py:656]: ok, worth parameterizing over dtype then? this is *just* testing a single EA type

Comment (jbrockmendel) [reply pandas/tests/frame/indexing/test_where.py:656]: not worth worrying about ATM, hopefully will get a solution before _too_ long that avoids object-dtype cast

**Review by jreback — APPROVED** [review]

**Pull request merged**